### DISCUSSION TITLE
Using wire:model saves input as model attributes and removes loaded relations

### DISCUSSION THREAD

**Discussion by kokoshneta** [livewire/livewire#7553] [1. Help] [answered]
[[ Laravel 9, Livewire 2 ]]

I have what I thought was rather a simple, foolproof component, representing a quote estimate. It has a public model property `$estimate`, which is mounted with three eager-loaded relations: one HasOne relation (`printQuote`) and two BelongsToMany relations (`externalQuotes` and `internalQuotes`). Some relation properties have `wire:model` input fields in the rendered component.

On initial load, everything works fine – all relations are loaded and available. Outputting `$this->estimate` in DebugBar before rendering gives these relations:

```php
#relations: array:3 [▼
  "printQuote" => App\Models\PrintQuote {#2112 ▶}
  "internalQuotes" => Illuminate\Database\Eloquent\Collection {#2108 ▼
    #items: array:3 [▼
      0 => App\Models\PriceQuote {#2254 ▶}
      1 => App\Models\PriceQuote {#2230 ▶}
      2 => App\Models\PriceQuote {#2229 ▶}
    ]
    #escapeWhenCastingToString: false
  }
  "externalQuotes" => Illuminate\Database\Eloquent\Collection {#2246 ▼
    #items: array:3 [▼
      0 => App\Models\PriceQuote {#2257 ▶}
      1 => App\Models\PriceQuote {#2213 ▶}
      2 => App\Models\PriceQuote {#2212 ▶}
    ]
    #escapeWhenCastingToString: false
  }
]
```

But if I change anything in any of the `wire:model`’ed input fields so Livewire makes a round trip to update the component, **the corresponding relationships are completely replaced with empty arrays** (not even collections, just plain arrays). Meanwhile, the actual input fields that I’d changed are saved as **model attributes**. The same DebugBar output as above now gives this:

```php
#attributes: array:18 [▼
  // other model attributes...
  "externalQuotes" => array:1 [▼
    0 => array:1 [▼
      "price" => "4"
    ]
  ]
]
#relations: array:3 [▼
  "printQuote" => App\Models\PrintQuote {#2196 ▶}
  "externalQuotes" => []
  "internalQuotes" => []
]
```

This is the relevant parts of the component class:

```php
class EstimateEditor extends Component {
  public Estimate $estimate;

  protected $rules = [
    'estimate.description' => 'string',
    'estimate.name' => 'string',
    'estimate.externalQuotes.*.price' => 'required|numeric',
    'estimate.internalQuotes.*.price' => 'required|numeric',
  ];

  public function mount(Estimate $estimate) {
    $this->estimate = $estimate->loadMissing(['printQuote', 'internalQuotes', 'externalQuotes']);
  }

  public function render() {
    dbg($this->estimate); // Output to DebugBar
    return view('livewire.estimates.show');
  }
}
```

The relationships in the `Estimate` model:

```php
public function externalQuotes() {
  return $this->quotes()->where('type', 1);
}

public function printQuote() {
  return $this->hasOne(PrintQuote::class);
}

public function internalQuotes() {
  return $this->quotes()->where('type', 0);
}

public function quotes() {
  return $this->hasMany(PriceQuote::class);
}
```

And an excerpt from the component view showing a modelled input field:

```blade
<div>
 @foreach ($estimate->externalQuotes as $quoteKey => $quote)
  <div>
   {{ $quote->description }}
  </div>
  <input type="text" wire:model.lazy="estimate.externalQuotes.{{ $quoteKey }}.price" />
 @endforeach
</div>
```

As far as I can tell from [the docs on binding to model properties and relations](https://laravel-livewire.com/docs/2.x/properties#binding-models), I’m doing everything by the book – but this behaviour is surely not intended.

### What might I be missing here?

**Comment by joshhanley** [comment, marked as answer]
@kokoshneta Livewire 2 has bugs when it comes to model binding nested relations. Specifically if you have any relations with multi-words then it won't work properly, such as your `externalQuotes` relation.

There is no work around for this, so much so that we have disabled model binding in V3 due to the amount of bugs and complications it has.

I would suggest refactoring your data properties to contain primitive data types (arrays and strings) or upgrade to V3 and make use of the new FormObjects.

Hope this helps!

**Reply by kokoshneta** [reply]
Aha – I would never have guessed that the camelCased relation names were the culprit! I renamed them `internals` and `externals`, and everything immediately worked. Thank you!

The lack of model data binding is in fact the primary reason I _haven’t_ yet upgraded to V3. The ability to directly bind to Eloquent model and relation properties is/was to me by far the most central and important feature Livewire offers/offered. I’ve yet to come across a situation in any of my projects where form objects would be useful as a substitute.
